x/reward/abci.go

```diff
@@ -9,60 +9,77 @@ import (
+	"github.com/axelarnetwork/utils/funcs"
 	"github.com/axelarnetwork/utils/slices"
 )
 
 // BeginBlocker is called at the beginning of every block
 func BeginBlocker(ctx sdk.Context, _ abci.RequestBeginBlock, _ types.Rewarder) {}
 
 // EndBlocker is called at the end of every block, process external chain voting inflation
-func EndBlocker(ctx sdk.Context, _ abci.RequestEndBlock, k types.Rewarder, n types.Nexus, m types.Minter, s types.Staker, msig types.MultiSig, ss types.Snapshotter) []abci.ValidatorUpdate {
-	handleExternalChainVotingInflation(ctx, k, n, m, s)
-	handleKeyMgmtInflation(ctx, k, m, s, msig, ss)
+func EndBlocker(ctx sdk.Context, _ abci.RequestEndBlock, k types.Rewarder, n types.Nexus, m types.Minter, s types.Staker, slasher types.Slasher, msig types.MultiSig, ss types.Snapshotter) []abci.ValidatorUpdate {
+	handleExternalChainVotingInflation(ctx, k, n, m, s, slasher, ss)
+	handleKeyMgmtInflation(ctx, k, m, s, slasher, msig, ss)
 
 	return nil
 }
 
-func addRewardsByConsensusPower(ctx sdk.Context, s types.Staker, rewardPool exported.RewardPool, validators []stakingtypes.Validator, totalReward sdk.DecCoin) {
+func addRewardsByConsensusPower(ctx sdk.Context, s types.Staker, rewardPool exported.RewardPool, validators []snapshot.ValidatorI, totalReward sdk.DecCoin) {
 	totalAmount := totalReward.Amount
 	denom := totalReward.Denom
 
-	validatorsWithConsensusPower := slices.Filter(validators, func(v stakingtypes.Validator) bool {
+	validatorsWithConsensusPower := slices.Filter(validators, func(v snapshot.ValidatorI) bool {
 		return v.GetConsensusPower(s.PowerReduction(ctx)) > 0
 	})
-	totalConsensusPower := slices.Reduce(validatorsWithConsensusPower, sdk.ZeroInt(), func(total sdk.Int, v stakingtypes.Validator) sdk.Int {
+	totalConsensusPower := slices.Reduce(validatorsWithConsensusPower, sdk.ZeroInt(), func(total sdk.Int, v snapshot.ValidatorI) sdk.Int {
 		return total.AddRaw(v.GetConsensusPower(s.PowerReduction(ctx)))
 	})
 
-	slices.ForEach(validatorsWithConsensusPower, func(v stakingtypes.Validator) {
+	slices.ForEach(validatorsWithConsensusPower, func(v snapshot.ValidatorI) {
 		// Each validator receives reward weighted by consensus power
 		amount := totalAmount.MulInt64(v.GetConsensusPower(s.PowerReduction(ctx))).QuoInt(totalConsensusPower).RoundInt()
 		rewardPool.AddReward(
 			v.GetOperator(),
 			sdk.NewCoin(denom, amount),
 		)
 	})
+}
+
+func excludeJailedOrTombstoned(ctx sdk.Context, slasher types.Slasher, snapshotter types.Snapshotter, v snapshot.ValidatorI) bool {
+	isTombstoned := func(v snapshot.ValidatorI) bool {
+		consAdd, err := v.GetConsAddr()
+		if err != nil {
+			return true
+		}
+
+		return slasher.IsTombstoned(ctx, consAdd)
+	}
+
+	isProxyActive := func(v snapshot.ValidatorI) bool {
+		_, isActive := snapshotter.GetProxy(ctx, v.GetOperator())
+
+		return isActive
+	}
 
+	filter := funcs.And(
+		funcs.Not(snapshot.ValidatorI.IsJailed),
+		funcs.Not(isTombstoned),
+		isProxyActive,
+	)
+
+	return !filter(v)
```

Comment:
```suggestion
	filter := funcs.Or(
		snapshot.ValidatorI.IsJailed,
		isTombstoned,
		funcs.Not(isProxyActive),
	)

	return filter(v)
```
and then also rename the function

Reply:
function name should still rename the same, since the return value doesn't change. Updated